File: check_names.rb
require 'csv'
require 'json'
require 'cgi'
require 'net/http'
require 'namae'

BASE_URL = 'https://pub.sandbox.orcid.org/v2.0/search/'
TOKEN = '' # your token here

$org_name = 'cincinnati'
$csv_file = 'names.csv'

$results = Array.new

def parse_name(input_name)
  names = Namae.parse input_name
  names[0]
end

def search_url(name)
  "#{BASE_URL}?q=given-names:#{CGI::escape(name.given)}"\
  "~+AND+family-name:#{name.family}&fq=affiliation-org-name:#{$org_name}"
end

def query_api(name)
  url = URI.parse(BASE_URL)
  req = Net::HTTP::Get.new(search_url(name),
          initheader = {
            'Authorization type' => 'Bearer',
            'Content-Type' => 'application/json',
            'Access token' => TOKEN
          })
  res = Net::HTTP.start(url.host, url.port, use_ssl: true) do |http|
    http.request(req)
  end
  res.body
end

def parse_csv
  CSV.foreach($csv_file, col_sep: ';').each do |row|
    puts "Checking name... #{row[0]}"
    response = JSON.parse(query_api(parse_name(row[0])))
    if response['result'].empty?
      $results << "#{row[0]}, --"
    else
      response['result'].each do |value|
        $results << "#{row[0]}, #{value['orcid-identifier']['uri']}"
      end
    end
  end
end

def write_results
  File.open('results.csv', 'w') do |file|
    file.syswrite $results.join "\n"
  end
end

parse_csv
write_results

File: check_names.py
from nameparser import HumanName
from urllib import parse
import requests
import csv
import json

token = '' # Your API Token Here
base_url = 'https://pub.sandbox.orcid.org/v2.0/search/'
org_name = 'cincinnati'
csv_file = 'names.csv'
output_file = 'results.csv'
results = []

def parse_csv():
  with open(csv_file, newline = '') as csvfile:
    name_reader = csv.reader(csvfile, delimiter = ';')
    for row in name_reader:
      print("Checking name... " + row[0])
      query_results = query_api(parse_name(row[0]))['result']
      if query_results:
        for query_result in query_results:
          results.append(f"{row[0]}, {query_result['orcid-identifier']['uri']}\n")
      else:
        results.append(f"{row[0]}, --\n")

def parse_name(input_name):
  name = HumanName(input_name)
  return name

def query_api(name):
  url = (f"{base_url}?q=given-names:{parse.quote_plus(name.first)}~+AND+family-name:"
         f"{parse.quote_plus(name.last)}&fq=affiliation-org-name:{org_name}")

  head = { 'Authorization type':'Bearer',
          'Content-Type':'application/json',
          'Access token':token }

  response = requests.get(url, headers = head)
  return response.json()

def write_csv():
  output = open(output_file, 'w')
  for result in results:
    output.write(result)
  output.close

parse_csv()
write_csv()
